test-data/unit/check-based-intersection.test

```diff
@@ -0,0 +1,340 @@
+from __future__ import annotations
+
+from typing import Any
+
+a: Any & str
+reveal_type(a)  # N: Revealed type is "Any & str"
+a = 1  # E: Incompatible types in assignment (expression has type "int", variable has type "Any & str")  [assignment]
+a = ""
+
+b: Any
+assert isinstance(b, str)
+reveal_type(b)  # N: Revealed type is "str"
+[builtins fixtures/tuple.pyi]
+
+
+[case testOverrideInIntersection]
+from __future__ import annotations
+
+class A:
+    def foo(self) -> int:
+        return True
+class B(A):
+    def foo(self) -> bool:
+        return True
+
+b: A & B
+reveal_type(b.foo())  # N: Revealed type is "bool"
+[builtins fixtures/tuple.pyi]
+
+
+[case testOverloaded]
+from __future__ import annotations
+from typing import overload
+class A:
+    @overload
+    def foo(self): ...
+    @overload
+    def foo(self, x: int): ...
+    def foo(self, x=1): ...
+class B:
+    def foo(self, x: str): ...
+
+x: A & B
+x.foo(None)  # E: No overload variant of "foo" of "__main__.A & __main__.B" matches argument type "None"  [call-overload] \
+             # N: Possible overload variants: \
+             # N:     def foo(self) -> None \
+             # N:     def foo(self, x: int) -> None \
+             # N:     def foo(self, x: str) -> None
+[builtins fixtures/tuple.pyi]
+
+
+[case testGenericFunction]
+from __future__ import annotations
+
+from typing import TypeVar
+
+T = TypeVar('T')
+U = TypeVar('U')
+
+def foo(x: T, y: U) -> T & U: pass
+a: int
+b: object
+reveal_type(foo(a, b))  # N: Revealed type is "int"
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionInferenceWithTypeVarValues]
+from __future__ import annotations
+from typing import TypeVar
+AnyStr = TypeVar('AnyStr', bytes, str)
+class A: pass
+strA: str & A
+def f(x: AnyStr & A, *a: AnyStr): pass
+f(strA)
+f(strA, 'bar')
+f(strA, b'bar') # E: Value of type variable "AnyStr" of "f" cannot be "str | bytes"  [type-var]
+[builtins fixtures/primitives.pyi]
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionTwoPassInference-skip]
+from __future__ import annotations
+from typing import TypeVar
+T = TypeVar('T')
+U = TypeVar('U')
+def j(x: list[T] & list[U], y: list[T]) -> list[U]: pass
+
+a = [1]
+b = ['b']
+# We could infer: Since list[str] <: list[T], we must have T = str.
+# Then since list[int] <: Union[list[str], list[U]], and list[int] is
+# not a subtype of list[str], we must have U = int.
+# This is not currently implemented.
+j(a, b)
+[builtins fixtures/list.pyi]
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionContext]
+from __future__ import annotations
+
+from typing import TypeVar
+
+class A: pass
+class B: pass
+
+T = TypeVar('T')
+def f() -> list[T]: pass
+d1: list[int] & A = f()  # E: Incompatible types in assignment (expression has type "list[int]", variable has type "list[int] & A")  [assignment]
+d2: A & B = f() # E: Incompatible types in assignment (expression has type "list[Never]", variable has type "A & B")  [assignment]
```

Comment:
do generics always get infered as `Never` when it can't figure out what they are? is that correct?

Reply:
https://mypy-play.net/?mypy=latest&python=3.11&gist=959e67bca3fb851d8d9f684d10bd9b43

yeah